token_count type : add enable_position_increments option (fix #23227)

fbaligand opened the pull request:
Add option "enable_position_increments" with default value true.
If option is set to false, indexed value is the number of tokens
(not position increments count)

Fix #23227

fbaligand committed:
token_count type : add an option to count tokens (fix #23227)

Add option "enable_position_increments" with default value true.
If option is set to false, indexed value is the number of tokens
(not position increments count)

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

jimczi reviewed (changes requested):
This looks great, thanks @fbaligand 
I left one comment regarding the meaning of the option. The title of the PR is misleading I think, `enable_position_increments` should not take 0-increment token into account so it's still the number of positions that we are after. We simply don't want to count positions where all tokens have been filtered.

Can you also add a description of the option in the docs:
https://github.com/elastic/elasticsearch/blob/master/docs/reference/mapping/types/token-count.asciidoc

fbaligand commented on core/src/test/java/org/elasticsearch/index/mapper/TokenCountFieldMapperTests.java:
```diff
@@ -102,6 +102,28 @@ public TokenStreamComponents createComponents(String fieldName) {
+                    return new TokenStreamComponents(new MockTokenizer(), tokenStream);
+                }
+            };
+        assertThat(TokenCountFieldMapper.countTokens(analyzer, "", ""), equalTo(3));
```
Hi @jimczi 
First, thanks for this first review.

Then, concerning your comment : the feature global goal is to have an option to count the number of tokens after analysis, whatever their positions. In that unit test, there are 3 tokens in the stream, so that's why I expect 3 as a result.

Given your comment, maybe the option "enable_position_increments" is misnamed for this feature goal ?

jimczi replied:
I think the naming is fine. This feature as described in the issue is about not counting tokens filtered from the token stream. This is what `enable_position_increments=false` does and I think it's all we need to do here. If your analyzer adds alternative tokens to each position they should not alter the final count since we're looking for the number of tokens in the original text.

fbaligand replied:
OK, I get what you mean.
That makes sense, especially when you use `synonym` filter.

=> I will fix all your requested changes and push them soon.

fbaligand committed:
fix enable_position_increments=false to count only distinct positions

fbaligand committed:
add documentation for new option 'enable_position_increments'

fbaligand committed:
remove compilation warning on TokenCount classes

fbaligand commented:
Hi @jimczi 

I fixed your requested changes.

By the way, I fixed also some compilation warnings on TokenCount classes.
If you want that I cancel this commit, tell me.

jimczi commented:
jenkins please test this

jimczi commented on docs/reference/mapping/types/token-count.asciidoc:
```diff
@@ -50,9 +50,10 @@ GET my_index/_search
 counting tokens. This means that even if the analyzer filters out stop
-words they are included in the count.
+words they are included in the count.  
+If you don't want to count stop words, set option `enable_position_increments` to `false`
```
this is redundant with the definition below and it's a little bit confusing since it impacts not only the stop filter ?

fbaligand replied:
You're right.
I see 2 options : 
- remove the note.
- only keep this part to just explain explain the mechanism :

> Technically, the `token_count` type sums position increments rather than counting tokens.

Which do you prefer ?

Moreover, I can add this precision to `enable_position_increments` definition :

> Set to `true` if you want to count all tokens computed by analyzer tokenizer. 

What do you think ?

jimczi replied:
I think you can just remove the note. The definition of the param below is descriptive enough.

fbaligand replied:
So, for enable_position_increments definition, you mean to not add "Set to true if you want to count all tokens computed by analyzer tokenizer" ?

jimczi replied:
Yep the current definition is ok.

jimczi reviewed (approved):
This is ready I think. I left one comment regarding the docs that could clarify the usage. 
This PR is actually tested by our internal CI so I'll merge it if it passes the test.
Thank you very much for your contribution @fbaligand !

fbaligand commented:
OK. I removed the note, rebased my code on master branch, and pushed it !

fbaligand commented:
Great !
Thanks @jimczi  for the review, merge into master and inclusion in 5.5.0 release !
It was a pleasure to "work" on this PR with you :)
